rel/graph.c

```diff
@@ -0,0 +1,217 @@
+  gDPPipeSync(NOW_LINE_OPA_DISP++);
+  gDPFullSync(NOW_LINE_OPA_DISP++);
+  gSPEndDisplayList(NOW_LINE_OPA_DISP++);
+
+  CLOSE_DISP(this);
+
+  if (*(u16*)GET_DYNAMIC_OFS(SYSDYNAMIC_HEADER_OFS) != SYSDYNAMIC_START_MAGIC) {
+    _dbg_hungup(__FILE__, 417);
+  }
+
+  if (*(u16*)GET_DYNAMIC_OFS(SYSDYNAMIC_EPILOG_OFS) != SYSDYNAMIC_END_MAGIC) {
+    _dbg_hungup(__FILE__, 425);
+    err = 1;
+  }
+
+  err = THA_GA_isCrash(&this->line_translucent_thaga) != 0;
+  err = THA_GA_isCrash(&this->overlay_thaga) != 0;
+  err = THA_GA_isCrash(&this->line_opaque_thaga) != 0;
+  err = THA_GA_isCrash(&this->polygon_opaque_thaga) != 0;
+  err = THA_GA_isCrash(&this->polygon_translucent_thaga) != 0;
+  err = THA_GA_isCrash(&this->font_thaga) != 0;
+  err = THA_GA_isCrash(&this->shadow_thaga) != 0;
+  err = THA_GA_isCrash(&this->light_thaga) != 0;
+
+  return err;
+}
+
+static void do_soft_reset(GAME* game) {
+  SoftResetEnable = FALSE;
+  //mBGM_reset();
+  //mVibctl_reset();
+  //sAdo_SoftReset();
+  ResetTime = osGetTime();
+  ResetStatus = IRQ_RESET_PRENMI;
+}
+
+static void reset_check(GRAPH* this, GAME* game) {
+  if (SoftResetEnable && osShutdown) {
+    do_soft_reset(game);
+  }
+}
+
+static void graph_main(GRAPH* this, GAME* game) {
+  game->disable_prenmi = FALSE;
+  graph_setup_double_buffer(this);
+  //game_get_controller(game);
+  game->disable_display = FALSE;
+  GRAPH_SET_DOING_POINT(this, GAME_MAIN);
+  game_main(game);
+  GRAPH_SET_DOING_POINT(this, GAME_MAIN_FINISHED);
+  if (ResetStatus < IRQ_RESET_DELAY) {
+    if (skip_frame < GETREG(SREG, 3)) {
+      skip_frame++;
+      this->frame_counter++;
+    }
+    else if (game->disable_display == FALSE) {
+      skip_frame = 0;
+      if (graph_draw_finish(this) == FALSE) {
+        GRAPH_SET_DOING_POINT(this, TASK_SET);
+        graph_task_set00(this);
+        GRAPH_SET_DOING_POINT(this, TASK_SET_FINISHED);
+        this->frame_counter++;
+
+        if ((GETREG(SREG, 33) & 1) != 0) {
+          SETREG(SREG, 33, GETREG(SREG, 33) & ~1);
+        }
+      }
+    }
+  }
+
+  if (GETREG(SREG, 20) < 2) {
+    GRAPH_SET_DOING_POINT(this, AUDIO);
+    //sAdo_GameFrame();
+    GRAPH_SET_DOING_POINT(this, AUDIO_FINISHED);
+  }
+
+  reset_check(this, game);
+
+  if (ResetStatus == IRQ_RESET_PRENMI && game->disable_prenmi == FALSE) {
+    //game->running = FALSE;
+    //game->next_game_init = prenmi_init;
+    //game->next_game_size = sizeof(PRENMI_GAME); /* sizeof(PRENIM_GAME) == 0xE8 */
+  }
+}
+
+extern void graph_proc(void* arg) {
+  int i;
+
+  DLFTBL_GAME* dlftbl = &game_dlftbls[0];
+  graph_ct(&graph_class);
+  
+  while (dlftbl != NULL) {
+    size_t size = dlftbl->alloc_size;
+    GAME* game = (GAME*)malloc(size);
+    game_class_p = game;
+    bzero(game, size);
+    GRAPH_SET_DOING_POINT(&graph_class, GAME_CT);
+    game_ct(game);
+    //emu64_refresh();
+    GRAPH_SET_DOING_POINT(&graph_class, GAME_CT_FINISHED);
+
+    for (i; i = game_is_doing(game); i != 0) {
```

Comment:
The first and last clauses in here shouldn't do anything, right? They're statements, not assignments, and aren't checked for truth. This seems more like a while loop, no `i`.

Reply:
This is just personal preference, but when a statement is assigned and checked as a loop condition, I favor using `for` loops. I dislike the syntax `while ((x = func()), x != y)`.